documentation/source/testbench_tools.rst

```diff
@@ -0,0 +1,189 @@
+
+                
+Busses
+======
+
+Busses are simply defined as collection of signals. The :py:class:`Bus` class will automatically bundle any group of signals together that are named similar to dut.<bus_name><separator><signal_name>. for instance,
+
+.. code-block:: python
+
+        dut.stream_in_valid
+        dut.stream_in_data
+        
+have a bus name of ``stream_in``, a separator of ``_``, and signal names of ``valid`` and ``data``. A list of signal names, or a dictionary mapping attribute names to signal names is also passed into the :py:class:`Bus` class. Busses can have values driven onto them, be captured (returning a dictionary), or sampled and stored into a similar object. 
+
+.. code-block:: python
+
+                stream_in_bus = Bus(dut, "stream_in", ["valid", "data"]) # '_' is the default separator
+
+
+Driving Busses
+==============
+
+examples and specific bus implementation bus drivers (amba, avalon, xgmii, and others) exist in the :py:class:`Driver` class enabling a test to append transactions to perform the serialization of transactions onto a physical interface. Here's an example using the avalon bus driver in the endian swapper example
+
+.. code-block:: python
+
+    class EndianSwapperTB(object):
+    
+        def __init__(self, dut, debug=False):
+            self.dut = dut
+            self.stream_in = AvalonSTDriver(dut, "stream_in", dut.clk)
+    
+    def run_test(dut, data_in=None, config_coroutine=None, idle_inserter=None,
+                 backpressure_inserter=None):
+    
+        cocotb.fork(Clock(dut.clk, 5000).start())
+        tb = EndianSwapperTB(dut)
+    
+        yield tb.reset()
+        dut.stream_out_ready <= 1
+    
+        if idle_inserter is not None:
+            tb.stream_in.set_valid_generator(idle_inserter())
+    
+        # Send in the packets
+        for transaction in data_in():
+            yield tb.stream_in.send(transaction)
+    
+
+Monitoring Busses
+=================
+
+For our testbenches to actually be useful, we have to monitor some of these busses, and not just drive them. That's where the :py:class:`Monitor` class comes in, with prebuilt Monitors for **avalon** and **xgmii** busses. The Monitor class is a base class which you are expected to derive for your particular purpose. You must create a `_monitor_recv()` function which is responsible for determining 1) at what points in simulation to call the `_recv()` function, and 2) what transaction values to pass to be stored in the monitors receiving queue. Monitors are good for both outputs of the dut for verification, and for the inputs of the dut, to drive a test model of the dut to be compared to the actual dut. For this purpose, input monitors will often have a callback function passed that is a model. This model will often generate expected transactions, which are then compared using the :py:class:`Scoreboard` class.
```

Comment:
A small suggestion: It's much easier to diff and edit large chunks of text if you stick to 1 sentence per line.

Reply:
Yeah, I noticed paying attention to diffs this is a bigger pain. I was trying to take advantage of the auto end of line in markdown, but based on this review I've cleaned up some of these long lines.